### PR TITLE
migration(tips): unreserve deposits

### PR DESCRIPTION
**Pull request by liamaharon**
Partial https://github.com/paritytech/polkadot-sdk/issues/485

## Overview

Adds a migration that unreserves all deposits made by the `tips` pallet.

## Details

`tips` contains a `Tips` storage item containing amounts reserved by open tips.

`pre_upgrade` records the reserves of all accounts with open tips prior to the migration, for use later in post_upgrade.

`on_runtime_upgrade` unreserves all balances locked up in tips.

`post_upgrade` checks that the reserved balances were reduced by the expected amount.

## Questions for reviewers / todos:

- [x] ~I'm not sure how to calculate the weights in `on_runtime_upgrade`, would appreciate some guidance there.~

### CODE REVIEW

File: frame/tips/src/migrations/unreserve_all_funds.rs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+		// Check that the reserved balance is reduced by the expected deposited amount.
+		for (account, actual_reserved_before) in account_reserved_before {
+			let actual_reserved_after = T::Currency::reserved_balance(&account);
+			let expected_amount_deducted = *account_deposits
+				.get(&account)
+				.expect("account deposit must exist to be in account_reserved_before, qed");
+			let expected_reserved_after =
+				actual_reserved_before.saturating_sub(expected_amount_deducted);
+			assert!(
```
Comment (ggwpez) [inline frame/tips/src/migrations/unreserve_all_funds.rs:160]: Not sure if `ensure!` can do formatting.

File: frame/tips/src/migrations/unreserve_all_funds.rs [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+		// Deposited funds need to be unreserved.
+		for (account, unreserve_amount) in account_deposits.iter() {
+			if unreserve_amount.is_zero() {
+				continue
+			}
+			T::Currency::unreserve(&account, *unreserve_amount);
+		}
+
+		RocksDbWeight::get()
+			.reads_writes(account_deposits.len() as u64, account_deposits.len() as u64) +
```
Comment (ggwpez) [inline frame/tips/src/migrations/unreserve_all_funds.rs:133]: `saturating_add` please.

File: frame/tips/src/migrations/unreserve_all_funds.rs [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+			account_deposits.iter().all(|(account, deposit)| *deposit <=
+				*account_reserved_before.get(account).unwrap_or(&Zero::zero())),
+			"Deposit amount is greater than reserved amount"
+		);
+
+		// Print some summary stats
+		let total_deposits_to_unreserve =
+			account_deposits.clone().into_values().sum::<BalanceOf<T, I>>();
+		log::info!("Total accounts: {}", account_deposits.keys().count());
+		log::info!("Total amount to unreserve: {:?}", total_deposits_to_unreserve);
```
Comment (ggwpez) [inline frame/tips/src/migrations/unreserve_all_funds.rs:112]: What stats does it report on Kusama and Polkadot?

**Review by ggwpez — APPROVED** [review]
Nice!

File: frame/tips/src/migrations/unreserve_all_funds.rs (hunk 3 above)
Comment (ggwpez) [reply frame/tips/src/migrations/unreserve_all_funds.rs:112]: Do you also have them for Polkadot? Just to gauge the values.

Comment (liamaharon) [reply frame/tips/src/migrations/unreserve_all_funds.rs:112]: Based on a snapshot taken ~1 week ago:
```
ElectionsPhragmen total impacted accounts: 323
ElectionsPhragmen stake to unlock: 240,352,972 DOT
ElectionsPhragment deposit to unreserve: 5,668 DOT

Democracy total impacted accounts: 5125
Democracy stake to unlock: 44,792,283 DOT
Democracy deposit to unreserve: 2,308 DOT

Tips total impacted accounts: 12
Tips amount to unreserve: 37 DOT
```

proportionally, a lot less than Kusama

File: frame/tips/src/migrations/unreserve_all_funds.rs (hunk 1 above)
Comment (liamaharon) [reply frame/tips/src/migrations/unreserve_all_funds.rs:161]: It does not do formatting. I can't seem to get `format!` working as a second arg, any advice?

Comment (ggwpez) [reply frame/tips/src/migrations/unreserve_all_funds.rs:161]: Hm nope… @kianenigma any idea?  
Maybe you can just log it.

File: frame/tips/src/migrations/unreserve_all_funds.rs [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+				// Add the balance to the account's existing deposit in the accumulator
+				acc.entry(tip.finder).or_insert(Zero::zero()).saturating_accrue(tip.deposit);
+				acc
+			});
+
+		(account_deposits, RocksDbWeight::get().reads(tips_len))
+	}
+}
+
+impl<T: crate::Config<I> + pallet_treasury::Config<I>, I: 'static> OnRuntimeUpgrade
```
Comment (muharem) [inline frame/tips/src/migrations/unreserve_all_funds.rs:68]: you can use a separate type parameter with a Currency trait bound to not depend from pallet_treasury here.

**Review by muharem — APPROVED** [review]

**Comment by muharem** [comment]
I wonder now if these migrations really needs to be part of the pallets. 
The pros, it can be reused, but probably very rarely. The cons, it should be maintained. The storage or logic change can make it wrong. 
Usually migrations are from one storage schema to another, which stays relevant.

**Comment by ggwpez** [comment]
I think we can delete pallet-tips from substrate once it is un-deployed from DotSama, so no maintenance burden.

File: frame/tips/src/migrations/unreserve_all_funds.rs (hunk 1 above)
Comment (liamaharon) [reply frame/tips/src/migrations/unreserve_all_funds.rs:161]: I'll log it for now. Should we also open an issue to add formatting support to `ensure!`?

**Comment by liamaharon** [comment]
> I think we can delete pallet-tips from Substrate once it is un-deployed from DotSama, so no maintenance burden.

@ggwpez are we sure there're no parachains using it?

**Comment by liamaharon** [comment]
bot merge

**Comment by ggwpez** [comment]
> > I think we can delete pallet-tips from Substrate once it is un-deployed from DotSama, so no maintenance burden.
> 
> @ggwpez are we sure there're no parachains using it?

Parachains can use it all they want, but maintaining pallets that we dont use ourselves does not make any sense (to me).  
So once we have a proper offboarding, probably containing a multi-block unlock migration, we should be able to delete them. Parachains can then fork the code and maintain if themselfes.

**Comment by liamaharon** [comment]
bot merge

**Comment by xlc** [comment]
> Parachains can use it all they want, but maintaining pallets that we dont use ourselves does not make any sense (to me).
So once we have a proper offboarding, probably containing a multi-block unlock migration, we should be able to delete them. Parachains can then fork the code and maintain if themselfes.

Really????

**Comment by ggwpez** [comment]
The Gov1 pallets can hopefully also be moved out of Substrate - to ORML for example.

**Comment by xlc** [comment]
I guess you should talk with ORML maintainers about it first before making a decision / suggestion.

**Comment by gilescope** [comment]
@xlc has a fair point. We could at least move old pallets to an unmaintained/archived dir for a period of time with the hope that others projects might take up the maintenance so that there's a period of time for people to migrate off it to something else. We need to be better at backward compatibility rather than rug pulling pallets out from under people. People can't build on shifting sands.

**Comment by gavofyork** [comment]
I have to mostly agree with @ggwpez here. Like any collective of individuals, Parity is neither able nor implicitly responsible to indefinitely maintain superseded or deprecated code, even if it was the one who originally wrote it.

**Comment by xlc** [comment]
I agree this should be a collective effort not just Parity but also there are some work that Parity have to do to make it possible. This isn’t a new topic. https://github.com/paritytech/substrate/discussions/11530

**Pull request merged**